t/rest_server/main_test.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/abcxyz/pkg/logging"
+	"github.com/abcxyz/pkg/renderer"
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestRealMain(t *testing.T) {
+	ctx := logging.WithLogger(context.Background(), logging.TestLogger(t))
+	ctx, done := context.WithCancel(ctx)
+	defer done()
+
+	var realMainErr error
+	finishedCh := make(chan struct{})
+	go func() {
+		defer close(finishedCh)
+		realMainErr = realMain(ctx)
+	}()
+
+	time.Sleep(100 * time.Millisecond) // wait for server startup
+	resp, err := http.Get(fmt.Sprintf("http://localhost:%s/", defaultPort))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer resp.Body.Close()
+
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	want := "hello world"
+	if !strings.Contains(string(b), want) {
+		t.Errorf("unexpected response: (-got,+want)\n%s", cmp.Diff(string(b), want))
+	}
+
+	// stop server
+	done()
+
+	// Wait for done
+	select {
+	case <-finishedCh:
+	case <-time.After(time.Second):
+		t.Fatalf("expected server to be stopped")
+	}
+
+	if realMainErr != nil {
+		t.Errorf("realMain(): %v", realMainErr)
+	}
+}
+
+func TestHandleHello(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	h := renderer.NewTesting(ctx, t, nil)
+
+	cases := []struct {
+		name string
+		want string
+	}{
+		{
+			name: "success",
+			want: "hello world",
+		},
+	}
+	for _, tc := range cases {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			server := httptest.NewServer(handleHello(h))
+			t.Cleanup(func() { server.Close() })
```

Comment:
I think we can just use `defer server.Close()` instead of t.Cleanup(). Both will work, but I think it's good to be consistent, and defer is a little simpler.

The main use case for t.Cleanup() is for helper functions that create test resources, and these can't use `defer` because the function returns before the test finishes.

Reply:
I generally always use `t.Cleanup` because I can explain the rule "if you're in _test.go, use t.Cleanup, otherwise use defer" pretty cleanly. However, "if you're in _test.go and the test isn't a parent test creating resources that need to live until all subtests complete, use t.Cleanup, otherwise use defer" always felt like more cognitive overhead to me.

Either way, let's pick something and put it in the style guide.

Reply:
This rule would be pretty simple: "use defer when you can, if not, use t.Cleanup()"

Reply:
But how does one know "when you can"? It feels like asking developers to take in a lot of context to make that decision.

Reply:
t.Cleanup() also means its easier to refactor duplicated code into a helper.

Reply:
I'm fine with t.Cleanup() since the consensus seems to be going that way.

Reply:
Ok, will leave as-is.